Sources/Apollo/InputValue.swift

```diff
@@ -0,0 +1,118 @@
+import Foundation
+
+/// For backwards compatibility with legacy codegen.
+/// The `GraphQLVariable` class has been replaced by `InputValue.variable`
+public func GraphQLVariable(_ name: String) -> InputValue {
+  return .variable(name)
+}
+
+/// Represents an input value to an argument on a `GraphQLField`'s `FieldArguments`.
+///
+/// - See: [GraphQLSpec - Input Values](http://spec.graphql.org/June2018/#sec-Input-Values)
+public indirect enum InputValue {
+  /// A direct input value, valid types are `String`, `Int` `Float` and `Bool`.
+  /// For enum input values, the enum cases's `rawValue` as a `String` should be used.
+  case scalar(JSONEncodable)
+
+  /// A variable input value to be evaluated using the operation's `variables` dictionary at runtime.
+  case variable(String)
+
+  /// A GraphQL "List" input value.
+  /// - See: [GraphQLSpec - Input Values - List Value](http://spec.graphql.org/June2018/#sec-List-Value)
+  case list([InputValue])
+
+  /// A GraphQL "InputObject" input value. Represented as a dictionary of input values.
+  /// - See: [GraphQLSpec - Input Values - Input Object Values](http://spec.graphql.org/June2018/#sec-Input-Object-Values)
+  case object([String: InputValue])
+
+  /// A null input value.
+  /// - See: [GraphQLSpec - Input Values - Null Value](http://spec.graphql.org/June2018/#sec-Null-Value)
+  case none
+
+  func evaluate(with variables: [String: JSONEncodable]?) throws -> JSONValue {
+    switch self {
+    case .scalar(let scalar):
+      return scalar.jsonValue
+
+    case .variable(let name):
+      guard let value = variables?[name] else {
+        throw GraphQLError("Variable \(name) was not provided.")
+      }
+      return value.jsonValue
+
+    case .list(let array):
+      return try evaluate(values: array, with: variables)
+
+    case .object(let dictionary):
+      return try evaluate(values: dictionary, with: variables)
+
+    case .none:
+      return NSNull()
+    }
+  }
+}
+
+extension InputValue: ExpressibleByNilLiteral {
+  @inlinable public init(nilLiteral: ()) {
+    self = .none
+  }
+}
+
+extension InputValue: ExpressibleByStringLiteral {
+  @inlinable public init(stringLiteral value: StringLiteralType) {
+    self = .scalar(value)
+  }
+}
+
+extension InputValue: ExpressibleByIntegerLiteral {
+  @inlinable public init(integerLiteral value: IntegerLiteralType) {
```

Comment:
what's the benefit of explicitly specifying this as `@inlineable`? In theory I thought the compiler inlined things like this anyway.

Reply:
Just for performance. You're right, the compiler will 99% do that on it's own, but this makes it more likely that it will get inlined.

Reply:
Does it sort of save the compiler the trouble of determining whether something can be inlined?

Reply:
Not 100% sure if it improves the build times or not, but I know it's a hint for the compiler that is should prefer inlining this whenever possible.